polkadot/node/core/pvf/src/artifacts.rs

```diff
@@ -251,21 +260,47 @@ impl Artifacts {
 		})
 	}
 
-	/// Remove artifacts older than the given TTL and return id and path of the removed ones.
-	pub fn prune(&mut self, artifact_ttl: Duration) -> Vec<(ArtifactId, PathBuf)> {
-		let now = SystemTime::now();
-
+	/// Remove artifacts older than the given TTL or the total artifacts size limit and return id
+	/// and path of the removed ones.
+	pub fn prune(&mut self, cleanup_by: &CleanupBy) -> Vec<(ArtifactId, PathBuf)> {
 		let mut to_remove = vec![];
-		for (k, v) in self.inner.iter() {
-			if let ArtifactState::Prepared { last_time_needed, ref path, .. } = *v {
-				if now
-					.duration_since(last_time_needed)
-					.map(|age| age > artifact_ttl)
-					.unwrap_or(false)
-				{
-					to_remove.push((k.clone(), path.clone()));
+
+		match cleanup_by {
+			CleanupBy::Time(artifact_ttl) => {
+				let now = SystemTime::now();
+				for (k, v) in self.inner.iter() {
+					if let ArtifactState::Prepared { last_time_needed, ref path, .. } = *v {
+						if now
+							.duration_since(last_time_needed)
+							.map(|age| age > *artifact_ttl)
+							.unwrap_or(false)
+						{
+							to_remove.push((k.clone(), path.clone()));
+						}
+					}
 				}
-			}
+			},
+			CleanupBy::Size(size_limit) => {
+				let mut total_size = 0;
+				let mut artifact_sizes = vec![];
+
+				for (k, v) in self.inner.iter() {
+					if let ArtifactState::Prepared { ref path, last_time_needed, .. } = *v {
+						if let Ok(metadata) = fs::metadata(path) {
```

Comment:
What bothers me here is that we're running a (possibly large) number of synchronous blocking filesystem operations in `Artifacts::prune()`, which is called from the PVF host's main loop running on a non-blocking threadpool. In case of I/O problems the whole PVF host will stall. I think we should either make use of `tokio::fs::metadata()`, or, even better, store the artifact's size as a property of the artifact itself along with the other data, and then no filesystem access is required inside `prune()`.

Reply:
Yeah, added the size to artifact's state